src/soc/rockchip/rk3399.c

```diff
@@ -0,0 +1,377 @@
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[2] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[2])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[3] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[3])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    if((rk3399->gpio[4] = (unsigned char *)mmap(0, rk3399->page_size, PROT_READ | PROT_WRITE, MAP_SHARED, rk3399->fd, rk3399->base_addr[4])) == NULL) {
+        wiringXLog(LOG_ERR, "wiringX failed to map the %s %s GPIO memory address", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+
+    return 0;
+}
+
+static char *rk3399GetPinName(int pin) {
+    return rk3399->layout[pin].name;
+}
+
+static void rk3399SetMap(int *map, size_t size) {
+    rk3399->map = map;
+    rk3399->map_size = size;
+}
+
+static void rk3399SetIRQ(int *irq, size_t size) {
+    rk3399->irq = irq;
+    rk3399->irq_size = size;
+}
+
+static int rk3399DigitalWrite(int i, enum digital_value_t value) {
+    struct layout_t *pin = NULL;
+    volatile unsigned int *data_reg = 0;
+
+    pin = &rk3399->layout[setPin(rk3399->map[i], rk3399->map)];
+
+    if(rk3399->map == NULL) {
+        wiringXLog(LOG_ERR, "the %s %s has not yet been mapped", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+    if(rk3399->fd <= 0 || rk3399->gpio == NULL) {
+        wiringXLog(LOG_ERR, "the %s %s has not yet been setup by wiringX", rk3399->brand, rk3399->chip);
+        return -1;
+    }
+    if(pin->mode != PINMODE_OUTPUT) {
+        wiringXLog(LOG_ERR, "the %s %s GPIO%d is not set to output mode", rk3399->brand, rk3399->chip, i);
+        return -1;
+    }
+
+    if(pin->bank == 1) {
+        data_reg = (volatile unsigned int *)(rk3399->gpio[1] + pin->data.offset);
+    } else if (pin->bank == 2) {
+        data_reg = (volatile unsigned int *)(rk3399->gpio[2] + pin->data.offset);
+    } else if (pin->bank == 3) {
+        data_reg = (volatile unsigned int *)(rk3399->gpio[3] + pin->data.offset);
+    } else if (pin->bank == 4) {
+        data_reg = (volatile unsigned int *)(rk3399->gpio[4] + pin->data.offset);
+    }
+
+    if(value == HIGH){
+        *data_reg |= (1 << (pin->bitNum));
+    }else if (value == LOW) {
+        *data_reg &= ~(1 << (pin->bitNum));
+    }
+
+    return 0;
+}
+
+static int rk3399DigitalRead(int i) {
+    struct layout_t *pin = NULL;
+    volatile unsigned int *data_reg = 0;
+    uint32_t val = 0;
+
+    pin = &rk3399->layout[setPin(rk3399->map[i], rk3399->map)];
```

Comment:
Missing validation

Reply:
I'm sorry to bother you that did the "Missing validation"  is the GPIO not verified ?

Reply:
I'm sorry to bother you that does the "Missing validation" is the GPIO not verified ?

Reply:
That you are not checking if the pins passed are within the proper range.